File: src/main/java/com/example/application/views/TodoView.java
package com.example.application.views;

import com.example.application.model.Todo;
import com.example.application.service.TodoService;
import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.html.Anchor;
import com.vaadin.flow.component.html.H1;
import com.vaadin.flow.component.html.ListItem;
import com.vaadin.flow.component.html.UnorderedList;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.component.textfield.TextField;
import com.vaadin.flow.data.binder.BeanValidationBinder;
import com.vaadin.flow.router.Route;

@Route("todo-view-server")
public class TodoView extends VerticalLayout {

  TodoView(TodoService service) {
    var task = new TextField("Task");
    var add = new Button("Add");
    var todoList = new UnorderedList();
    var form = new HorizontalLayout(task, add);
    form.setAlignItems(Alignment.BASELINE);

    add(new H1("Todos"), form, todoList, new Anchor("todo-view-client", "Client-side view"));

    for (Todo todo : service.getTodos()) {
      todoList.add(new ListItem(todo.getTask()));
    }

    var binder = new BeanValidationBinder<>(Todo.class);
    binder.forField(task).bind("task");

    add.addClickListener(e -> {
      var todo = new Todo();
      if (binder.writeBeanIfValid(todo)) {
        var saved = service.saveTodo(todo);
        todoList.add(new ListItem(saved.getTask()));
        binder.readBean(new Todo());
      }
    });
  }
}

File: frontend/views/todo-view.ts
import { LitElement, customElement, html, property, css } from "lit-element";
import Todo from "../generated/com/example/application/model/Todo";
import { getTodos, saveTodo } from "../generated/TodoService";
import { field, Binder } from "@vaadin/form";
import TodoModel from "../generated/com/example/application/model/TodoModel";
import "@vaadin/vaadin-text-field";
import "@vaadin/vaadin-button";

@customElement("todo-view")
export class TodoView extends LitElement {
  @property({ type: Array })
  private todos: Todo[] = [];
  private binder = new Binder(this, TodoModel);

  static styles = css`
    :host {
      margin: var(--lumo-space-m);
      display: block;
    }

    .form {
      display: inline-grid;
      gap: var(--lumo-space-m);
      grid-template-columns: 3fr 1fr;
      align-items: baseline;
    }
  `;

  protected render() {
    return html`
      <h1>Todos</h1>
      <div class="form">
        <vaadin-text-field
          label="Task"
          ...=${field(this.binder.model.task)}
        ></vaadin-text-field>
        <vaadin-button @click=${this.add}>Add</vaadin-button>
      </div>
      <ul>
        ${this.todos.map((todo) => html` <li>${todo.task}</li> `)}
      </ul>

      <a href="todo-view-server" router-link>Server-side view</a>
    `;
  }

  async firstUpdated() {
    this.todos = await getTodos();
  }

  async add() {
    const saved = await this.binder.submitTo(saveTodo);
    if (saved) {
      this.todos = [...this.todos, saved];
      this.binder.clear();
    }
  }
}
